File: intellij-community/platform/lang-impl/src/com/intellij/util/indexing/diagnostic/IndexingJobStatistics.java
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.util.indexing.diagnostic;

import com.intellij.openapi.fileTypes.FileType;
import org.jetbrains.annotations.NotNull;

import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

public class IndexingJobStatistics {
  final ConcurrentMap<String, PerThreadTime> timesPerIndexer = new ConcurrentHashMap<>();
  final ConcurrentMap<String, PerThreadTime> timesPerFileType = new ConcurrentHashMap<>();
  final ConcurrentMap<String, Integer> numberOfFilesPerFileType = new ConcurrentHashMap<>();
  final PerThreadTime contentLoadingTime = new PerThreadTime();
  final PerThreadTime indexingTime = new PerThreadTime();

  public void addFileStatistics(@NotNull FileIndexingStatistics fileStatistics, @NotNull FileType fileType) {
    fileStatistics.perIndexerTimes.forEach((indexId, time) -> {
      timesPerFileType
        .computeIfAbsent(indexId.getName(), (__) -> new PerThreadTime())
        .addTimeSpentInCurrentThread(time);
    });
    String fileTypeName = fileType.getName();
    numberOfFilesPerFileType.compute(fileTypeName, (__, currentNumber) -> (currentNumber != null ? currentNumber : 0) + 1);
    timesPerFileType
      .computeIfAbsent(fileTypeName, (__) -> new PerThreadTime())
      .addTimeSpentInCurrentThread(fileStatistics.totalTime);
  }

  public void addIndexingTime(long ms) {
    indexingTime.addTimeSpentInCurrentThread(ms);
  }

  public void addContentLoadingTime(long ms) {
    contentLoadingTime.addTimeSpentInCurrentThread(ms);
  }
}

File: intellij-community/platform/lang-impl/src/com/intellij/util/indexing/diagnostic/IndexingJobStatistics.kt
// Copyright 2000-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
package com.intellij.util.indexing.diagnostic

import java.util.concurrent.atomic.AtomicInteger

class IndexingJobStatistics {

  private val _statsPerIndexer = hashMapOf<String, StatsPerIndexer>()
  val statsPerIndexer: Map<String /* ID.name() */, StatsPerIndexer>
    @Synchronized get() = _statsPerIndexer.toMap()

  private val _statsPerFileType = hashMapOf<String, StatsPerFileType>()
  val statsPerFileType: Map<String /* File type name */, StatsPerFileType>
    @Synchronized get() = _statsPerFileType.toMap()

  val numberOfTooLargeForIndexingFiles = AtomicInteger()
  val tooLargeForIndexingFiles = LimitedPriorityQueue<TooLargeForIndexingFile>(5, compareBy { it.fileSize })

  data class StatsPerIndexer(
    val indexingTime: TimeStats,
    var numberOfFiles: Int,
    var totalBytes: BytesNumber
  )

  data class StatsPerFileType(
    val indexingTime: TimeStats,
    val contentLoadingTime: TimeStats,
    var numberOfFiles: Int,
    var totalBytes: BytesNumber
  )

  @Synchronized
  fun addFileStatistics(
    fileStatistics: FileIndexingStatistics,
    contentLoadingTime: Long,
    fileSize: Long
  ) {
    fileStatistics.perIndexerTimes.forEach { (indexId, time) ->
      val stats = _statsPerIndexer.getOrPut(indexId.name) {
        StatsPerIndexer(TimeStats(), 0, 0)
      }
      stats.indexingTime.addTime(time)
      stats.numberOfFiles++
      stats.totalBytes += fileSize
    }
    val fileTypeName = fileStatistics.fileType.name
    val stats = _statsPerFileType.getOrPut(fileTypeName) {
      StatsPerFileType(TimeStats(), TimeStats(), 0, 0)
    }
    stats.contentLoadingTime.addTime(contentLoadingTime)
    stats.indexingTime.addTime(fileStatistics.indexingTime)
    stats.totalBytes += fileSize
    stats.numberOfFiles++
  }
}